types/module/module.go

```diff
@@ -298,7 +299,16 @@ func (m *Manager) InitGenesis(ctx sdk.Context, cdc codec.JSONCodec, genesisData
 		}
 		ctx.Logger().Debug("running initialization for module", "module", moduleName)
 
-		moduleValUpdates := m.Modules[moduleName].InitGenesis(ctx, cdc, genesisData[moduleName])
+		moduleGenesis := m.Modules[moduleName].DefaultGenesis()
+		if moduleGenesis == nil {
+			continue
+		}
```

Comment:
This seems wrong to me, we shouldn't be calling `DefaultGenesis` inside the module manager's InitGenesis.

Reply:
Thanks for this comment, I was quite unsure about this change. What I need here is some way to get the genesis type of the module so then we can unmarshal the `json.RawMessage` into it. Do you think we should have a method like `GenesisZeroMessage()` that returns an empty genesis struct?

Reply:
Wait, why are we using default genesis state here? InitGenesis should use whatever state was provided in the genesis file.

Reply:
@alexanderbez see here, I'm using it to get the right `proto.Message` so then I can unmarshal the genesis file json into it:

https://github.com/cosmos/cosmos-sdk/blob/d304999d478c775e707f76162172c453e62c8d02/types/module/module.go#L302-L309

Reply:
Ahh I see. This feels a lil dirty to me lol, but it makes sense!

Reply:
Just found a bug here... We'll need a method that returns an empty genesis struct instead of the default genesis because there's no overwriting happening.
So if your genesis file contains a value that equals to 0 and the default genesis has a non-zero value, then you end up merging the genesis from the file with the default genesis.

```go
// default genesis
mockAppModule1.EXPECT().DefaultGenesis().Return(&testdata.Cat{Lives: 123}

// genesis json
"module1": json.RawMessage(`{"moniker": "module1"}`),

// Results in:
 Got: moniker:"module1" lives:123  (*testdata.Cat)
 Want: is equal to moniker:"module1"  (*testdata.Cat)

```

`EmptyGenesis()` sounds ok?

Reply:
I don't like adding a `EmptyGenesis()` to AppModule, seems a bit useless.

Is there a way to use reflection? Something like:
```go
f := reflect.TypeOf(m.Modules[moduleName].DefaultGenesis)
if f.Kind() != reflect.Func {
	panic("not a function...")
}
genesisType := f.Out(0) // Get 1st return value's type
moduleGenesis := reflect.New(genesisType).Elem().Interface().(proto.Message)
```

(might need some playing around to make this work).

Reply:
If we are ok with reflection there, let's do it. I'll give it a go 👌

Reply:
Funny as that was gonna be my suggestion -- use reflection or introduce a new API. I have a preference on the former.

Reply:
This is what I came up with. I need to execute the DefaultGenesis function to get the actual type, if I just try to get the returning type I always get proto.Message

https://github.com/cosmos/cosmos-sdk/blob/0181ccf013734e9bc27e9a91fb18886067d1a8c2/types/module/module.go#L302-L310